### PR TITLE
d3.fsum

### LINKED ISSUE #152: Implement exact summation (fsum)?
Discussed with Kahan summation (#33, #149), fsum is a bit slower but provably "exact" and would be the preferred "exact summation" algorithm for d3-array.

I've implemented it in https://observablehq.com/d/723090bd2ef54f5f ; it does not return a sum but an object that can be used to continue the sum by adding more terms.

- [x] speed test (about 2x slower than d3.sum which is not that bad)
- [x] a valueof accessor
- [x] check if this implementation is useful in d3-geo.
- [x] document

### PR DESCRIPTION
**Pull request by Fil**
exact summation of IEEE 754 numbers
fixes #152 (see also #33)

### CODE REVIEW

**Comment by mbostock** [comment]
d3.fsum should just return a number. And it should require an iterable as input, as d3.sum does.

If we want something else for iterative adding, we should expose that as a separate interface, such as a d3.adder class. And I think we should simplify that class substantially: it should just have an add(number) method, no iterables, no accessor function. This class would then be used internally by d3.fsum, and if people want the more expressive interface they can use d3.adder instead of d3.fsum.

**Comment by Fil** [comment]
I've fixed the API and made the code more readable.

File: README.md [hunk 1]
```diff
@@ -130,6 +130,29 @@ Returns an [unbiased estimator of the population variance](http://mathworld.wolf
...
+
+Although slower, d3.fsum can replace d3.sum everywhere an exact summation is needed. Uses <a href="#adder">d3.Adder</a>.
+
+<a name="adder" href="#adder">#</a> new d3.<b>Adder</b>()
+
+Creates a full precision adder for [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754) floating point numbers, setting its value to 0.
+
+<a name="adder_add" href="#adder_add">#</a> *adder*.<b>add</b>(number)
+
+Adds *number* to the adder’s current value.
```
Comment (mbostock) [inline README.md:150]:
```suggestion
Adds *number* to the adder’s current value and returns the adder.
```

File: src/fsum.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
+// adapted from https://github.com/python/cpython/blob/master/Modules/mathmodule.c#L1444
+export class Adder {
+  constructor() {
+    this.partials = new Float64Array(32).fill(0);
+    this.n = 0;
```
Comment (mbostock) [inline src/fsum.js:5]: These should probably have underscores to indicate that they are internal variables.

```suggestion
    this._partials = new Float64Array(32).fill(0);
    this._n = 0;
```

File: src/fsum.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
+// adapted from https://github.com/python/cpython/blob/master/Modules/mathmodule.c#L1444
```
Comment (mbostock) [inline src/fsum.js:1]:
```suggestion
// https://github.com/python/cpython/blob/a74eea238f5baba15797e2e8b570d153bc8690a7/Modules/mathmodule.c#L1423
```

**Pull request merged**